src/jshint.js

```diff
@@ -2748,6 +2761,219 @@ var JSHINT = (function() {
+    if (state.tokens.next.value === "extends") {
+      advance("extends");
+      expression(0, state.tokens.next); //?
+    }
+
+    if (state.tokens.next.value === "{") {
+      advance("{");
+    } else {
+      warning("W116", state.tokens.curr, "identifier", state.tokens.next.type); //?
+      advance();
+    }
+
+    return classNameToken;
+  }
+
+  // Class statement
+  blockstmt("class", function(context) {
+    var className, classNameToken;
+    var inexport = context & prodParams.export;
+
+    classNameToken = classHeader();
+    if (classNameToken) {
+      className = classNameToken.value;
+    }
+
+    if (className) {
+      this.name = className;
+      state.funct["(scope)"].addlabel(className, {
+        type: "class",
+        initialized: true,
+        token: classNameToken
+      });
+      if (inexport) {
+        state.funct["(scope)"].setExported(className, classNameToken);
+      }
+    }
+    state.funct["(scope)"].stack();
+
+    classBody.call(this, context);
+    return this;
+  }).exps = true;
+
+  // Class expression
+  prefix("class", function(context) {
+    var className, classNameToken;
+
+    classNameToken = classHeader();
+    if (classNameToken) {
+      className = classNameToken.value;
+    }
+
+    //TODO: the constructor's 'name' should be the name of the variable in an assignment expression, ie in 'var x = class Y {}' I believe it's supposed to be 'x'.
+
+    // In a class Expression: the name should not be saved into the calling scope, but is still accessible inside the definition.
+    // So we open a new scope first, then save the name.
+    state.funct["(scope)"].stack();
+    if (className) {
+      this.name = className;
+      state.funct["(scope)"].addlabel(className, {
+        type: "class",
+        initialized: true,
+        token: classNameToken
+      });
+      state.funct["(scope)"].block.use(className, classNameToken);
+    }
+
+    classBody.call(this, context);
+    return this;
+  });
+
+  function classBody(context) {
+    var props = {};
+    var name, accessorType, token, isStatic, inGenerator;
+
+    // TODO: this should probably be an update to state.funct, but I'm not sure the right way to do it.
+    state.inClassBody = true;
+    // This is one way to make super() and super.method() parse. It might not be the best (could interfere with metrics or something)
+    state.funct["(scope)"].block.add("super", "function", null, false, true);
+
+    while (state.tokens.next.value !== "}") {
+      isStatic = false;
+      inGenerator = false;
+      if (state.tokens.next.value === "static") {
+        isStatic = true;
+        advance();
+      }
+      if (state.tokens.next.value === "*") {
+        inGenerator = true;
+        advance();
+      }
+
+      token = state.tokens.next;
+      switch (token.value) {
+        case ";":
+          warning("W032", token);
+          advance();
+          break;
+        case "constructor":
+          if (isStatic || inGenerator) {
```

Comment:
Actually, static methods may be named "constructor", so (confusing though it may be) we'll need to allow that.